server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancingRoundSummary.java

```diff
@@ -9,16 +9,129 @@
+     * balance. Each summary represents a step towards the next desired balance, which is based on presuming the previous desired balance is
+     * reached. So combining them is roughly the difference between the first summary's previous desired balance and the last summary's
+     * latest desired balance.
+     *
+     * @param numberOfBalancingRounds How many balancing round summaries are combined in this report.
+     * @param nodeNameToWeightChanges
+     * @param numberOfShardMoves The sum of shard moves for each balancing round being combined into a single summary.
+     */
+    public record CombinedBalancingRoundSummary(
+        int numberOfBalancingRounds,
+        Map<String, NodesWeightsChanges> nodeNameToWeightChanges,
+        long numberOfShardMoves
+    ) {
+
+        public static final CombinedBalancingRoundSummary EMPTY_RESULTS = new CombinedBalancingRoundSummary(0, new HashMap<>(), 0);
+
+        public static CombinedBalancingRoundSummary combine(List<BalancingRoundSummary> summaries) {
+            if (summaries.isEmpty()) {
+                return EMPTY_RESULTS;
+            }
+
+            // Initialize the combined weight changes with the oldest changes. We can then build the combined changes by adding the diffs of
+            // newer weight changes. If a new node gets added in a later summary, then we will initialize its weights starting there.
+            // Similarly, a node may be removed in a later summary: in this case we will keep that nodes work, up until it was removed.
+            var iterator = summaries.iterator();
+            assert iterator.hasNext();
+            var firstSummary = iterator.next();
+            Map<String, NodesWeightsChanges> combinedNodeNameToWeightChanges = new HashMap<>(firstSummary.nodeNameToWeightChanges);
+
+            // Number of shards moves are simply summed across summaries. Each new balancing round is built upon the last one, so it is
+            // possible that a shard is reassigned back to a node before it even moves away, and that will still be counted as 2 moves here.
+            long numberOfShardMoves = firstSummary.numberOfShardsToMove;
+
+            // Initialize with 1 because we've already begun to iterate the summaries.
+            int numSummaries = 1;
+
+            // Iterate any remaining summaries (after the first one).
+            while (iterator.hasNext()) {
+                var summary = iterator.next();
+                for (var nodeNameAndWeights : summary.nodeNameToWeightChanges.entrySet()) {
+                    var combined = combinedNodeNameToWeightChanges.get(nodeNameAndWeights.getKey());
+                    if (combined == null) {
+                        // Encountered a new node in a later summary. Add the new node initializing it with the base weights from that
+                        // summary.
+                        combinedNodeNameToWeightChanges.put(nodeNameAndWeights.getKey(), nodeNameAndWeights.getValue());
```

Comment:
I realized that this is essentially the same case as the first summary round initialization I was doing in the `combinedNodeNameToWeightChanges` declaration. So I got rid of the complicated first iterator value handling I was doing before the while-loop, and simplified it all into the while-loop.